DSTORE-1475 out of order compliance operations

johnmbw opened the pull request:
This is intended to make it possible to support out of order compliance (and potentially metric update) messages.  I've made it fairly general so it's potentially possible to get this upstreamed in the future.

The rules are configured for the document update processor like:

```
         <lst name="existingPermanentDeletes">
            <str name="must">OLD.compliance_reason:delete</str>
            <str name="action">skip</str>
          </lst>
          <lst name="existingSoftDeletes">
            <str name="must">OLD.compliance_reason:soft_delete</str>
            <str name="action">upsert:compliance_reason</str>
          </lst>
          <lst name="newSoftDeletes">
            <str name="must">NEW.compliance_reason:soft_delete</str>
            <str name="action">upsert:*</str>
          </lst>
```

There can be multiple conditions and they are checked in order - the first one that matches wins.

The matching of clauses can be defined in a similar way to lucene:

* `must` - required to match
* `must_not` - required to not match
* `should` - at least one is required match

The clauses themselves look like:

* `OLD.my_field:the_value` - the document already in solr has a field called `my_field` with the value `the_value`
* `OLD.my_field:*` - the document already in solr has a field called `my_field` with any value
* `OLD.*` - the document already in solr exists
* `NEW.my_field:the_value` - the document about to be written has a field called `my_field` with the value `the_value`

The actions can look like:

* `skip` - don't write the new document at all
* `insert` - carry on as normal and write the document
* `upsert:*` - merge the document with the existing document in solr (if there is one) - copying over all fields not on the new document
* `upsert:my_field` - merge the document with the existing document in solr (if there is one) - copying over just `my_field`
* `upsert:my_field1,myfield2` - merge the document with the existing document in solr (if there is one) - copying over just `my_field1` & `my_field2`

johnmbw committed:
DSTORE-1475 add support for conditional upserts

Adds a document processor factory that can be used to conditionall
upsert or skip documents based on the values of specific fields in
either the new or the old documents.

This is mostly intended for handling compliance where the system
handling compliance will not have the original documents available, but
needs to flag documents that may or may not have been already written in
solr.

Example config:

      <lst name="existingPermanentDeletes">
        <str name="must">OLD.compliance_reason:delete</str>
        <str name="action">skip</str>
      </lst>
      <lst name="existingSoftDeletes">
        <str name="must">OLD.compliance_reason:soft_delete</str>
        <str name="action">upsert:compliance_reason</str>
      </lst>
      <lst name="newSoftDeletes">
        <str name="must">NEW.compliance_reason:soft_delete</str>
        <str name="action">upsert:*</str>
      </lst>

johnmbw committed:
DSTORE-1475 support conditions for fields simply existing

So we can unconditionally skip documents if one already exists in solr
for example.

johnmbw committed:
DSTORE-1475 add insert action for conditional upsert processor

This is so we can enable default behaviour under some circumstances

johnmbw committed:
DSTORE-1475 bit less verbose creation of namedlists in tests

johnmbw committed:
DSTORE-1475 support cases where old doc isn't present better

Also extracted more code into UpsertCondition so it can be tested

johnmbw committed:
DSTORE-1475 support conditions to check for document existing

johnmbw commented:
Another more complex example is:
```
<processor class="org.apache.solr.update.processor.ConditionalUpsertProcessorFactory">
      <!-- if we have a previous "tombstone" from a permanent delete this should stop new inserts -->
      <lst name="existingPermanentDeletes">
        <str name="must">OLD.compliance_reason:delete</str>
        <str name="action">skip</str>
      </lst>
      <!-- if the document has already been previously soft deleted we need to maintain that status -->
      <lst name="existingSoftDeletes">
        <str name="must">OLD.compliance_reason:soft_delete</str>
        <str name="action">upsert:compliance_reason</str>
      </lst>
      <!-- any non-compliance updates that are missing dates (partial updates) should be skipped as per the old SkipExistingDocumentsProcessorFactory  -->
      <lst name="otherPartialUpdates">
        <str name="must_not">OLD.*</str>
        <str name="must_not">NEW.bw_timestamp:*</str>
        <str name="must_not">NEW.compliance_reason:*</str>
        <str name="action">skip</str>
      </lst>
    </processor>
```

Which we can use to replicate the old `SkipExistingDocumentsProcessorFactory` to stick writes for partial documents when there isn't already a doc in solr.

spmason commented on solr/core/src/java/org/apache/solr/update/processor/UpsertCondition.java:
```diff
@@ -0,0 +1,276 @@
+  static boolean shouldInsertOrUpsert(List<UpsertCondition> conditions, SolrInputDocument oldDoc, SolrInputDocument newDoc) {
+    for (UpsertCondition condition: conditions) {
+      if (condition.matches(oldDoc, newDoc)) {
+        log.info("Condition {} matched, taking action", condition.getName());
```
Should these be DEBUG level instead?

johnmbw replied:
Damn - yep, meant to change those.

spmason reviewed (commented):
Have you got any indication of what the performance of this is like?

I suppose filtering out the partial documents on the querying side is an separate exercise?

johnmbw commented:
Haven't done any major performance testing yet, but I expect it shouldn't be much worse than the current skip/check if exists stuff, as that has to do the lookup first too.

Though yeah I guess if this is slower than looking up in HBase we should do that instead.  I think using this or hbase shouldn't matter too much for the rest of the compliance work though as that's mostly concerned with when events arrive _after_ the docs anyway (so regular partial updates etc will work fine there).

I think if we ensure the partial docs are missing the date (`bw_timestamp`) field and if we always filter to include a date range then filtering them out _should_ sort of happen naturally.

timatbw commented:
Still reviewing, but in general if we're now overwriting inserts always and we can achieve the same effect of skipping atomic-updates for missing docs, then we can use this in place of the SkipExisting processor completely, rather than pay the cost of doing both in sequence.

timatbw commented:
Another thing, is it harder to allow overriding behaviour per-request (like SkipExisting could do, because there were only 2 flags)? I guess you can define 2 or more chains and then pick by name in the request (the choice of chain to use is a parameter too)

timatbw commented on solr/core/src/java/org/apache/solr/update/processor/ConditionalUpsertProcessorFactory.java:
```diff
@@ -0,0 +1,123 @@
+
+  @Override
+  public void init(NamedList args)  {
+    conditions.clear();
```
Do you think init could get called twice? It should only create one instance per solr core and never reuse the object (I hope!)

johnmbw replied:
Yeah just being defensive really.  If there had a been a `setAll` method I would have used that instead.

timatbw commented on solr/core/src/java/org/apache/solr/update/processor/ConditionalUpsertProcessorFactory.java:
```diff
@@ -0,0 +1,123 @@
+      }
+
+      if (distribProc == null) {
+        throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "DistributedUpdateProcessor must follow DocComplianceUpdateProcessor");
```
message needs updating

timatbw commented on solr/core/src/java/org/apache/solr/update/processor/ConditionalUpsertProcessorFactory.java:
```diff
@@ -0,0 +1,123 @@
+  static class ConditionalUpsertUpdateProcessor extends UpdateRequestProcessor {
+    private final SolrCore core;
+
+    private DistributedUpdateProcessor distribProc;  // the distributed update processor following us
```
Feels odd that the distrib update should be _after_ this processor.. really this processor can only do its work on an actual index (i.e. once the request has arrived at the leader of the relevant shard)

timatbw commented on solr/core/src/java/org/apache/solr/update/processor/UpsertCondition.java:
```diff
@@ -0,0 +1,276 @@
+    }
+
+    private static Predicate<SolrInputDocument> forField(String field, Predicate<Object> fieldPredicate) {
+      return doc -> doc != null && fieldPredicate.test(doc.getFieldValue(field));
```
Not sure how this will work for non-String fields and multi-valued fields. `getFieldValue` returns an Object which could be an Integer etc. You might have to toString it first? And maybe also use `getFieldValues` for multi-valued fields.

johnmbw replied:
I was only thinking about single valued string fields, but I should probably make it handle multi-valued and non-string fields too.  There's no reason not to really.

timatbw commented on solr/core/src/test/org/apache/solr/update/processor/UpsertConditionTest.java:
```diff
@@ -0,0 +1,617 @@
+    assertThat(UpsertCondition.shouldInsertOrUpsert(conditions, oldDoc, newDoc), is(true));
+    assertThat(newDoc.getFieldValue("metric1"), is("kept-from-old1"));
+    assertThat(newDoc.getFieldValue("metric2"), is("kept-from-old2"));
+    assertThat(newDoc.getFieldValue("metric3"), nullValue());
```
I was wondering if it's possible to clear/remove a field from the NEW doc, given the OLD doc and conditions? Upsert looks like it can only ever append or change fields.

johnmbw replied:
Yeah I'd assumed we'd only want to append fields for upsert.  I guess it wouldn't be impossible to make it always use what's in the old doc, but I assumed we'd want a way to be able to actual edit fields.  Otherwise we'd have to delete the doc first always.

Could maybe just be a difference syntax to denote when the OLD value should always replace the NEW value.

timatbw replied:
yeah when we first chatted about this, I thought there would be 4 ways to decide how to handle a field: 1. firstnotnull(new, old) 2. firstnotnull(old, new) 3. always old, 4. always new

But at that point, we weren't really considering much beyond this compliance use-case and the Instagram use-case of improving but not downgrading docs.

timatbw commented:
Also trying to think how this interacts with atomic updates? Or does it only handle inserts which are possibly modified using the OLD document (which may have been created using an atomic update originally, but that's not relevant now).

If an atomic update comes through first, there's no existing doc, and what gets created is a sparse document containing only the fields that the atomic updates sets? Later when a full document insert comes through, this old sparse doc is what's used to augment the new doc, and that is _inserted_? i.e. this processor will not make changes to atomic update documents themselves?

johnmbw commented:
Yeah my plan was we'd have a partial/atomic update that would create the sparse doc, then that would be combined with the new (full) doc later.

So for a "soft delete" scenario we could send through a partial update:
* if there was already a doc it'd simply set the flag to the right value
* if the doc wasn't there it'd set the flag, then when the doc did arrive we'd upsert to ensure it stayed set

but yeah effectively there's no "special" handling of partial docs/atomic update.  I'd wondered about adding some syntax to match when we have a partial doc update, but I think usually we can rely on the absence of certain fields to achieve the same effect.

johnmbw commented:
> Another thing, is it harder to allow overriding behaviour per-request (like SkipExisting could do, because there were only 2 flags)? I guess you can define 2 or more chains and then pick by name in the request (the choice of chain to use is a parameter too)

I'd sort of assumed it'd be tricky to do, but I guess supporting multiple chains or maybe a flag to entirely skip the chain might be useful.

johnmbw committed:
DSTORE-1475 reduce log level to debug

johnmbw committed:
DSTORE-1475 support matching multi-values fields

johnmbw committed:
DSTORE-1475 handle non-string values when matching

johnmbw committed:
DSTORE-1475 add param so we can ignore conditional upsert logic entirely

In cases where we need to reset data in some fashion for example.

johnmbw committed:
DSTORE-1475 support action that will always keep OLD field values

This can be handy in situations where we want to "latch" a value in some
way.

johnmbw committed:
DSTORE-1475 nullify action for upsert processor

This action makes it possible to ensure certain fields are always forced
to be null.  e.g. if there are fields that should always be redacted if
some condition is met.

johnmbw committed:
DSTORE-1475 support multiple actions per condition

Also change things so conditions can fall through, which means
we can combine things more effectively.

johnmbw committed:
DSTORE-1475 no-op when copying fields and OLD doc does not exist

johnmbw commented:
Ok, so hopefully last round of additions for this now.  Main changes:

* Only `insert` and `skip` actions stop checking of later conditions now
* Support multiple actions per condition (which means you can decide if you want to stop further processing by adding an `insert` after an `upsert`)
* Add a `nullify` action which will force fields to be set to `null` when written to solr (for cases where we need to ensure only specific fields are redacted)

timatbw reviewed (approved):
Had a look at the most recent 3 commits, LGTM
